Web content process gets killed when loading apartments.com

szewai opened the pull request:
#### e6357e81a0403dcf89bf6b3048ce308c7450f2c6
<pre>
Web content process gets killed when loading apartments.com
<a href="https://bugs.webkit.org/show_bug.cgi?id=285046">https://bugs.webkit.org/show_bug.cgi?id=285046</a>
<a href="https://rdar.apple.com/141734000">rdar://141734000</a>

Reviewed by Ben Nham.

Web content process gets killed for sending `NotificationManagerMessageHandler` messages and failing message check (that
`NetworkProcess::builtInNotificationsEnabled()` must return true) in `NetworkConnectionToWebProcess::dispatchMessage()`.
The reason `NetworkProcess::builtInNotificationsEnabled()` returns false is UI process does not send correct value to
network process. In current implementation, UI process sets `NetworkProcess::builtInNotificationsEnabled()` only on
network process initialization (see `NetworkProcessProxy::sendCreationParametersToNewProcess`). The value in parameters
is decided based on `DeprecatedGlobalSettings::builtInNotificationsEnabled()` in UI process (see
`NetworkProcessProxy::sendCreationParametersToNewProcess`). `DeprecatedGlobalSettings::builtInNotificationsEnabled()` is
updated when `builtInNotificationsEnabled` preferences value is changed (see `WebPreferences::updateBoolValueForKey`).
However, if the default preferences value is true, `WebPreferences::updateBoolValueForKey()` is not invoked, so the
settings value will stay false. This means settings value (which network process uses) does not match preferences value
(which web content process uses).

Given that network process also tracks preferences in `NetworkConnectionToWebProcess::m_sharedPreferencesForWebProcess`,
we could just use that for message check, and this will fix the crash (as web process and network process have the same
`builtInNotificationsEnabled` value).

Regarding the problem that settings value does not match preferences value in UI process, this patch just replaces the
usage of settings value with preferences value. Instead of maintaining a global settings value and updating it when some
preferences object in UI process has change, we just compute a shared value based on all preferences (see
`WebsiteDataStore::builtInNotificationsEnabled()`). Also, UI process could include the value in message sent to network
process, so that network process does not need to maintain the state.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::dispatchMessage):
(WebKit::NetworkConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
(WebKit::NetworkProcess::processPushMessage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::updateBoolValueForKey):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::builtInNotificationsEnabled const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: <a href="https://commits.webkit.org/288384@main">https://commits.webkit.org/288384@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1601f09ccf335b2088cbe26ac0e8eb7041153998

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82221 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1885 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36280 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87353 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33282 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1956 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9765 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64079 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21823 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85290 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1350 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74811 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44357 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1250 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32323 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72552 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29614 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88709 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9527 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6782 "Found 1 new test failure: http/wpt/mediarecorder/record-96KHz-sources.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72470 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9753 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70626 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71688 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17930 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15822 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14846 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/879 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9480 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14991 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9354 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12820 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11123 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

bnham commented on Source/WebKit/UIProcess/WebPreferences.cpp:
```diff
@@ -197,11 +197,6 @@ void WebPreferences::updateBoolValueForKey(const String& key, bool value, bool e
 
-#if ENABLE(WEB_PUSH_NOTIFICATIONS)
-    if (key == WebPreferencesKey::builtInNotificationsEnabledKey())
-        WebCore::DeprecatedGlobalSettings::setBuiltInNotificationsEnabled(value);
```
How does this work exactly? Aren't there still a bunch of places where we're checking `DeprecatedGlobalSettings::builtInNotificationsEnabled()`?

bnham replied:
Hm ok, I see now that [288101@main](https://commits.webkit.org/288101@main) changed the implementation of `DeprecatedGlobalSettings::builtInNotificationsEnabled()`.

bnham reviewed (approved):
Let's land this now to fix the crash and discuss how to clean this all up next week with @beidson.

szewai committed:
Web content process gets killed when loading apartments.com
https://bugs.webkit.org/show_bug.cgi?id=285046
rdar://141734000

Reviewed by Ben Nham.

Web content process gets killed for sending `NotificationManagerMessageHandler` messages and failing message check (that
`NetworkProcess::builtInNotificationsEnabled()` must return true) in `NetworkConnectionToWebProcess::dispatchMessage()`.
The reason `NetworkProcess::builtInNotificationsEnabled()` returns false is UI process does not send correct value to
network process. In current implementation, UI process sets `NetworkProcess::builtInNotificationsEnabled()` only on
network process initialization (see `NetworkProcessProxy::sendCreationParametersToNewProcess`). The value in parameters
is decided based on `DeprecatedGlobalSettings::builtInNotificationsEnabled()` in UI process (see
`NetworkProcessProxy::sendCreationParametersToNewProcess`). `DeprecatedGlobalSettings::builtInNotificationsEnabled()` is
updated when `builtInNotificationsEnabled` preferences value is changed (see `WebPreferences::updateBoolValueForKey`).
However, if the default preferences value is true, `WebPreferences::updateBoolValueForKey()` is not invoked, so the
settings value will stay false. This means settings value (which network process uses) does not match preferences value
(which web content process uses).

Given that network process also tracks preferences in `NetworkConnectionToWebProcess::m_sharedPreferencesForWebProcess`,
we could just use that for message check, and this will fix the crash (as web process and network process have the same
`builtInNotificationsEnabled` value).

Regarding the problem that settings value does not match preferences value in UI process, this patch just replaces the
usage of settings value with preferences value. Instead of maintaining a global settings value and updating it when some
preferences object in UI process has change, we just compute a shared value based on all preferences (see
`WebsiteDataStore::builtInNotificationsEnabled()`). Also, UI process could include the value in message sent to network
process, so that network process does not need to maintain the state.

* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::dispatchMessage):
(WebKit::NetworkConnectionToWebProcess::dispatchSyncMessage):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
(WebKit::NetworkProcess::processPushMessage):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
(WebKit::NetworkProcessProxy::processPushMessage):
* Source/WebKit/UIProcess/WebPreferences.cpp:
(WebKit::WebPreferences::updateBoolValueForKey):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::builtInNotificationsEnabled const):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:

Canonical link: https://commits.webkit.org/288384@main
